drivers/gpio/gpio_mcp23x17.c

```diff
@@ -102,14 +118,34 @@ static int write_port_regs(const struct device *dev, uint8_t reg,
 {
 	struct mcp23x17_drv_data *const drv_data =
 		(struct mcp23x17_drv_data *const)dev->data;
-	int ret;
+	const struct mcp23x17_config *const config = dev->config;
+	int ret = -1;
 	uint16_t port_data;
 
-	LOG_DBG("MCP23X17: Write: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
-		reg, (value & 0xFF), (reg + 1), (value >> 8));
-
 	port_data = sys_cpu_to_le16(value);
 
+#if MCP23X17_BUS_I2C
+	const struct device *const i2c_master = drv_data->i2c_master;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+
+	int retry = 3;
```

Comment:
Can you explain why you need this to be in a loop?  This seems unnecessary.  If the chip can ignore I2C operations because it's busy, there ought to be a status you can check while waiting, rather than writing blindly.

Reply:
During testing I've seen situations where the i2c_burst_write() fails for some reason.
It might be due to bus-conditions as the board I'm testing it on has several other I2C-slaves and the software is heavily threaded. For the same reason I've added mutexes to the MCUX I2C-driver as these for some reason has been removed in later releases (not a discussion we're going to take here ;) ).

The datasheet states:
"... If a Stop or Restart condition is generated
during a data transfer, the data will not be written to the
MCP23017."
I'm not sure if this is the cause though as the function just returns a general IO-error and it happens only sporadically.
My costumer is paranoid of errors logged to the debug-shell which is why I added the loop instead of handling it at a higher level.

Anyway - if for compliancy reasons you want me to remove it I'll do it.
I'll just have a variant in my own tree.